MM-34665: Pricing differentiation features

agarciamontoro opened the pull request:
#### Summary
This PR finishes the first batch of pricing differentiation features. Namely:

1. It opens up the API and UI to all self-managed servers with any of E0, E10 or E20 licenses.
  - Cloud servers are untouched, as the flow to upgrade and the packaging features are still to be decided. So the behaviour should not change at all in cloud servers.
  - Team Edition servers have the UI still restricted (not the API, because differentiating between E0 and TE happens not to be possible as of today. 
2. It keeps three restrictions for the E0 users, who are not able to:
  - create more than one playbook per team,
  - access the Timeline tab in the RHS nor
  - add a post to the timeline through the post menu "Add to timeline" entry.
3. When interacting with these restricted features, the flow for admins and normal users is different:
  - As an admin, you get a modal (or, in the case of the RHS timeline, a placeholder image) from where you can start a trial.
  - As a normal user, you get a modal (or, in the case of the RHS timeline, a placeholder image) from where you can notify the administrator. This sends a DM to all the admins in the server, from where they can start the trial.

I encourage you to review this commit by commit. Each comment should be self-explanatory on its own, and the commit message gives some context on what it is doing. Also, all commits are independent (except for the couple ones that touch the `PricingPlanDifferentiationEnabled` flags, which are added in one commit and modified later), so you should not worry about reviewing edited code.

### Demos

#### As a normal user


https://user-images.githubusercontent.com/3924815/117140450-b73b5b00-adad-11eb-9d7a-f099f82f8d92.mp4


https://user-images.githubusercontent.com/3924815/117140457-ba364b80-adad-11eb-9e25-1f1eba166441.mp4


https://user-images.githubusercontent.com/3924815/117140467-bc98a580-adad-11eb-8fc6-95ccc8acb14f.mp4



#### As an admin:

https://user-images.githubusercontent.com/3924815/117140269-80654500-adad-11eb-81e4-f5a77a3c868e.mp4


https://user-images.githubusercontent.com/3924815/117140288-865b2600-adad-11eb-8796-fc9159aeb2b3.mp4


https://user-images.githubusercontent.com/3924815/117140302-89eead00-adad-11eb-9022-0e9a23275037.mp4

#### As an admin with notifications initiated by a normal user

https://user-images.githubusercontent.com/3924815/117140411-ac80c600-adad-11eb-9be4-d173e9c4e270.mp4

The notifications initiated from the Timeline tab and from the post menu entry are the same, so I'm showing just one here:

https://user-images.githubusercontent.com/3924815/117140432-b1de1080-adad-11eb-9378-047f1d04e9e6.mp4





#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34665
https://mattermost.atlassian.net/browse/MM-34666
https://mattermost.atlassian.net/browse/MM-34667

justinegeffen commented:
Thanks, @agarciamontoro! I haven't started reviewing but this is an awesome PR! Looking forward to going through it.

abhijit-singh reviewed (changes requested):
Great work @agarciamontoro!

A few small changes:

1.  In the normal user flow where they have an option to Notify System Admin, can you update the illustration and text on the success modal to match https://www.figma.com/file/1M0Kg4QpQ64fhHslLNeXK7/MM-34664-Pricing-Gate-UI?node-id=48%3A233031?
2. In the normal user flow modals, there does seem to be some extra space above and below the buttons - 
<img width="755" alt="extra-space" src="https://user-images.githubusercontent.com/14369607/117144941-2de95000-add0-11eb-9f3f-7751065d0637.png">
The space between the text and the buttons is correct in the Start trial modal - 
<img width="749" alt="correct-spacing" src="https://user-images.githubusercontent.com/14369607/117150415-c2a27c80-add5-11eb-9d9d-8c06c5534c85.png">

In the design, I had 32px between the text and the buttons, & 48 px below the buttons. 


3. In the timeline tab on the RHS, is it easily possible to hide the `Overview` & `Update status` buttons? They're conflicting with `Start trial` button because of proximity and aren't tightly related to the timeline feature.

crspeller commented on webapp/src/components/backstage/upgrade_modal.tsx:
```diff
@@ -0,0 +1,343 @@
+    }
+};
+
+const UpgradeModal: FC<Props> = (props: Props) => {
```
This is a complicated modal!

cpoile replied:
@crspeller Would it be better to have the different modals as separate components? (even if it meant some duplication)

agarciamontoro replied:
> This is a complicated modal!

I agree.

> Would it be better to have the different modals as separate components? (even if it meant some duplication)

I will try something similar to this, I'll ping you when it's ready.

agarciamontoro replied:
I ended up splitting it into smaller, simpler pieces, and I think it is now a bit more easy to digest. See https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/a34f311f2e6919d8c22c50bc3adae78be118a7d7. What do you think?

crspeller replied:
Much better!

crspeller commented on webapp/src/hooks.ts:
```diff
@@ -214,3 +217,45 @@ export function useLatestUpdate(incident: Incident) {
+
+    useEffect(() => {
+        const fetch = async () => {
+            const response = await clientFetchPlaybooksCount(team.id);
```
This is a lot of extra requests to grab the number of playbooks.

agarciamontoro replied:
This will need to change when we go to a workspace-scoped product, so it will be easier to change this there. I created https://mattermost.atlassian.net/browse/MM-35609 to track it.

crspeller commented on webapp/src/hooks.ts:
```diff
@@ -214,3 +217,45 @@ export function useLatestUpdate(incident: Incident) {
+
+function useIsAllowedInE0() {
+    const numPlaybooks = useNumPlaybooksInCurrentTeam();
+    const isLicensed = useSelector(isE10LicensedOrDevelopment);
```
`useIsAllowedInE0` -> `isE10Licenced` Confusing naming.

agarciamontoro replied:
We should keep the use because it is actually a hook (https://reactjs.org/docs/hooks-custom.html#using-a-custom-hook), but I agree the name is confusing, but it does not only check if the server is licensed, but if the creation of playbooks is allowed. What about being super explicit and `useIsPlaybookCreationAllowed`?

agarciamontoro replied:
Simplified this in https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/5b43013d93fc63d34c548df8ea03409b0fe85bd4

crspeller commented on webapp/src/hooks.ts:
```diff
@@ -214,3 +217,45 @@ export function useLatestUpdate(incident: Incident) {
+// useAllowTimelineViewInCurrentTeam returns whether a user can view the RHS
+// timeline in the current team
+export function useAllowTimelineViewInCurrentTeam() {
+    return useIsAllowedInE0();
```
So we allow timeline in E0? Why are we checking then?

agarciamontoro replied:
It is actually incorrect. This is the restriction for the playbooks, but the timeline cannot be viewed in E0 in any case. Thanks for catching this, I'm fixing it!

agarciamontoro replied:
Fixed in https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/5b43013d93fc63d34c548df8ea03409b0fe85bd4

crspeller commented on webapp/src/make_update_main_menu.ts:
```diff
@@ -9,16 +9,20 @@ import {getCurrentTeam} from 'mattermost-redux/selectors/entities/teams';
+        let show = !isMobile();
+
+        if (!isPricingPlanDifferentiationEnabled(store.getState())) {
+            const disable = isDisabledOnCurrentTeam(store.getState());
```
The disabling on the current team thing shouldn't be contingent on pricing plan differentiation being enabled. Right? Still confused by the `isPricingPlanDifferentiationEnabled` stuff.

agarciamontoro replied:
Oops, good catch! I will fix this logic.

crspeller reviewed (changes requested):
Lots of complicated stuff!

I am really confused by the `PricingPlanDifferentiationEnabled` stuff. Maybe because I don't understand what behavior that is trying to model?

cpoile commented on webapp/src/components/assets/buttons.tsx:
```diff
@@ -103,3 +105,20 @@ export const GrayTertiaryButton = styled.button`
 
+export type UpgradeButtonProps = React.ComponentProps<typeof PrimaryButton>;
+
+export const UpgradeButton: FC<UpgradeButtonProps> = (props: UpgradeButtonProps) => {
+    const {children, ...rest} = props;
+    return (
+        <PrimaryButton {...rest}>
+            {children}
+            <PositionedUpgradeBadge/>
+        </PrimaryButton>
```
This is cool!

cpoile reviewed (changes requested):
Awesome work @agarciamontoro ! This is a huge (and complex) feature, nicely done.

justinegeffen reviewed (commented):
This is awesome, @agarciamontoro! I've made some suggestions to terminology and also suggested changing "Administrator" to "System Admin" for consistency with our docs. Other than that, it looks good to me!

agarciamontoro commented:
@abhijit-singh 

> In the normal user flow where they have an option to Notify System Admin, can you update the illustration and text on the success modal to match https://www.figma.com/file/1M0Kg4QpQ64fhHslLNeXK7/MM-34664-Pricing-Gate-UI?node-id=48%3A233031?

Done:

![image](https://user-images.githubusercontent.com/3924815/117675589-fce09500-b1ac-11eb-8a36-a5d2c835079e.png)


> In the normal user flow modals, there does seem to be some extra space above and below the buttons

Fixed:

![image](https://user-images.githubusercontent.com/3924815/117675523-ef2b0f80-b1ac-11eb-94b3-c4ef2e92e4c6.png)

> In the timeline tab on the RHS, is it easily possible to hide the Overview & Update status buttons? They're conflicting with Start trial button because of proximity and aren't tightly related to the timeline feature.

Removed:

![image](https://user-images.githubusercontent.com/3924815/117675643-0cf87480-b1ad-11eb-9336-5d4d00f02662.png)

I also fixed what we discussed in the meeting last time:

1. I removed the purchase license button

![image](https://user-images.githubusercontent.com/3924815/117675762-28637f80-b1ad-11eb-9d35-09191324fad3.png)

2. I added a badge to the templates header

![image](https://user-images.githubusercontent.com/3924815/117675859-40d39a00-b1ad-11eb-99e0-9f5aa88daa79.png)

3. I modified the behaviour of the interactive message sent to the admin to change only the text of the button while the license is being downloaded:


https://user-images.githubusercontent.com/3924815/117676211-9d36b980-b1ad-11eb-9fe7-1f66172c2001.mp4

agarciamontoro commented on server/api/api.go:
```diff
@@ -36,11 +36,13 @@ func NewHandler(pluginAPI *pluginapi.Client, config config.Service) *Handler {
 
 	settingsRouter := handler.APIRouter.PathPrefix("/settings").Subrouter()
-	settingsRouter.Use(e20Middleware.Middleware)
+	if !config.IsPricingPlanDifferentiationEnabled() {
```
> I don't get this. If we are on cloud... which is what this function checks. We are defiantly licensed why do we do a check here?

The `IsPricingPlanDifferentiationEnabled` is actually a feature flag. The pricing plan differentiation is enabled only on on-prem servers (on the webapp side, it is further restricted to only on-prem servers **with at least E0**), because the upgrade flow for cloud is not yet finished. For now, all cloud servers are licensed, but this will change when 5.36 is released, which is the deadline for the completion of this feature (this PR and the one coming).

But now that I think about it, I can actually get rid of the whole thing on the server side, and simply have every endpoint open for all servers, licensed or not (when the feature is complete, this will be the case, so we can just do it now). The frontend will still be restricted while I implement the cloud upgrade flow. What do you think, @crspeller, @cpoile?

> I also think the IsPricingPlanDifferentiationEnabled name is confusing because why would you want to disable it? It seems pretty fundamental.

I agree that it is not my best naming decision :joy: We want to disable it in cloud because it is not yet implemented. Again, we can simply keep the flag in the client, and simply remove this one.

> Also it seems like this could be handled centrally by using a use on the api subrouter.

I did not want to add it as a subrouter because of the nature of the function being a feature flag. But yup, getting rid of it in the API would solve all of these cases.

agarciamontoro replied:
Done in https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/6b8b0026c53e51cb318d8948071f97559f3695a8

agarciamontoro commented on server/bot/poster.go:
```diff
@@ -82,6 +83,81 @@ func (b *Bot) PublishWebsocketEventToUser(event string, payload interface{}, use
+	admins, err := b.pluginAPI.User.List(&model.UserGetOptions{
+		Role:    string(model.SYSTEM_ADMIN_ROLE_ID),
+		Page:    0,
+		PerPage: 100,
```
I was paranoid too. We can bump it to 1000 no problem (what if there are more?!).

agarciamontoro replied:
Admin level over 1000 https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/db0d22888c780085795067d189b5987af9ea6102

agarciamontoro commented on server/command/command.go:
```diff
@@ -1398,7 +1398,7 @@ func (r *Runner) Execute() error {
 	}
 
-	if !r.configService.IsLicensed() {
+	if !config.PricingPlanDifferentiationEnabled && !r.configService.IsLicensed() {
```
If it is enabled, everything should be open. So when I get rid of the `PricingPlanDifferentiationEnabled`, I will simply get rid of this `if` altogether.

agarciamontoro commented on webapp/src/hooks.ts:
```diff
@@ -214,3 +217,45 @@ export function useLatestUpdate(incident: Incident) {
+    return numPlaybooks;
+}
+
+function useIsAllowedInE0() {
```
It is a hook, so according to https://reactjs.org/docs/hooks-custom.html#using-a-custom-hook it should be prefixed with `use`, right?

agarciamontoro commented on webapp/src/hooks.ts:
```diff
@@ -214,3 +217,45 @@ export function useLatestUpdate(incident: Incident) {
+
+// useAllowAddMessageToTimelineInCurrentTeam returns whether a user can add a
+// post to the timeline in the current team
+export function useAllowAddMessageToTimelineInCurrentTeam() {
+    return useIsAllowedInE0();
```
Yup, you're right, thank you!

agarciamontoro replied:
Fixed in https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/5b43013d93fc63d34c548df8ea03409b0fe85bd4

agarciamontoro commented on webapp/src/components/backstage/upgrade_modal.tsx:
```diff
@@ -0,0 +1,343 @@
+
+    const UserStartModal = Object.assign({...CommonModal}, {
+        helpText: helpText + ' Notify your administrator to upgrade.',
+        confirmButtonText: 'Notify Administrator',
```
@justinegeffen, should this be "Notify System Admin" as you suggested [for the RHS placehodler](https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/files#r629328206)? Or change that one to be simply "Notify", too. Otherwise, it will not be consistent.

justinegeffen replied:
Good question! Notify System Admin works for me, thanks. :)

agarciamontoro replied:
@justinegeffen, for both the "Add more to your timeline" and the "Playbook limit reached" modals, yes.

agarciamontoro commented on webapp/src/components/rhs/rhs_timeline_placeholder.tsx:
```diff
@@ -0,0 +1,185 @@
+
+    let helpText = 'Make retros easy. Your timeline includes all the events in your incident, separated by type, and downloadable for offline review.';
+    if (actionState === ActionState.Success) {
+        helpText = 'A notification has been sent to your administrator.';
```
@justinegeffen, same question here as before

cpoile commented on server/config/service.go:
```diff
@@ -183,12 +183,12 @@ func (c *ServiceImpl) IsCloud() bool {
-func (c *ServiceImpl) IsPricingPlanDifferentiationEnabled() bool {
-	// Disable for all cloud servers
-	return !c.IsCloud()
+// IsE10Licensed returns true when the server is appropriately licensed to run this plugin.
```
These two comments could be better I think -- reading just the `IsE20Licensed` comment makes me think we need E20 to run the plugin, but then this comment makes me think we need E10, so I'm left thinking "what licence do we need"?

agarciamontoro replied:
You're right! Fixed in https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/c9b7f9b4016e96ba39777775a9544a66453801b4

cpoile commented on webapp/src/hooks.ts:
```diff
@@ -235,27 +235,41 @@ export function useNumPlaybooksInCurrentTeam() {
+    return useSelector(isE10LicensedOrDevelopment);
+}
+
+export function useEnsureProfiles(userIds: string[]) {
```
Is this duplicating on purpose, or a mistake with github?

agarciamontoro replied:
Mmm not sure what you mean. Is something duplicated here? I don't see it :thinking:

cpoile reviewed (approved):
Cool, looks much nicer, thanks @agarciamontoro! That upgrade modal is kinda amazing. :) Nothing blocking, good work.

crspeller commented on server/api/bot.go:
```diff
@@ -0,0 +1,131 @@
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type BotHandler struct {
```
Why is this called bot handler? How about LicensingHandler?

agarciamontoro replied:
The root route is `/bot/notify-admins`, and I thought it we could use it for more bot-related activities, if needed?

crspeller commented on server/config/config.go:
```diff
@@ -31,6 +31,9 @@ type Service interface {
+	IsE20Licensed() bool
+
+	// IsE10Licensed returns true when the server is appropriately licensed to run this plugin.
+	IsE10Licensed() bool
```
I assume this will also succeed if the server has an E20 license correct? Maybe we should make this explicit in the name. `IsAtLeastE10Licensed`? At least change the comment to clarify.

agarciamontoro replied:
You're right, changed the comment in https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/546/commits/c06ed2e4745379b03a13182fbeea2c09f674cdbd. Thanks!

crspeller reviewed (approved):
Couple of non-blocking suggestions.

agarciamontoro commented:
/update-branch

agarciamontoro commented:
As discussed offline, QA will test this directly on master. Merging now! :tada:
